src/platform/Linux/NetworkCommissioningDriver.h

```diff
@@ -0,0 +1,166 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <platform/NetworkCommissioning.h>
+#include <vector>
+
+namespace chip {
+namespace DeviceLayer {
+namespace NetworkCommissioning {
+
+template <typename T>
+class LinuxScanResponseIterator : public Iterator<T>
+{
+public:
+    LinuxScanResponseIterator(std::vector<T> * apScanResponse) : mpScanResponse(apScanResponse) {}
+    size_t Count() override { return mpScanResponse != nullptr ? mpScanResponse->size() : 0; }
+    bool Next(T & item) override
+    {
+        if (mpScanResponse == nullptr || currentIterating >= mpScanResponse->size())
+        {
+            return false;
+        }
+        item = (*mpScanResponse)[currentIterating];
+        currentIterating++;
+        return true;
+    }
+    void Release() override
+    { /* nothing to do, we don't hold the ownership of the vector. */
+    }
```

Comment:
This is to actually release the memory of the iterator. Since the iterator may be long-lived and is passed by pointer, and you don't know how it was allocated, you cannot call `delete`/`free` on the pointer when you are done with the iterator after processing it in the callback. Therefore, this impl should release `this` somehow. See how it's done in the `GroupDataProviderImpl`